cmd/access/upstream_connector.go

```diff
@@ -0,0 +1,61 @@
+package main
+
+import (
+	"context"
+	"time"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/lifecycle"
+	"github.com/onflow/flow-go/network/p2p"
+)
+
+// upstreamConnector connects the unstaked AN with an upstream staked AN
+type upstreamConnector struct {
+	lm               *lifecycle.LifecycleManager
+	stakedANIdentity *flow.Identity
+	logger           zerolog.Logger
+	unstakedNode     *p2p.Node
+	cancel           context.CancelFunc
+}
+
+func newUpstreamConnector(stakedAN *flow.Identity, unstakedNode *p2p.Node, logger zerolog.Logger) *upstreamConnector {
+	return &upstreamConnector{
+		lm:               lifecycle.NewLifecycleManager(),
+		stakedANIdentity: stakedAN,
+		unstakedNode:     unstakedNode,
+		logger:           logger,
+	}
+}
+func (connector *upstreamConnector) Ready() <-chan struct{} {
+	connector.lm.OnStart(func() {
+		const maxAttempts = 3
+		ctx, cancel := context.WithCancel(context.Background())
+		connector.cancel = cancel
+		log := connector.logger.With().Str("staked_an", connector.stakedANIdentity.String()).Logger()
+		for i := 1; i <= maxAttempts; i++ {
+			err := connector.unstakedNode.AddPeer(ctx, *connector.stakedANIdentity)
+			if err == nil {
+				log.Info().Msg("connected to upstream access node")
+				return
+			}
+			log.Error().Int("attempt", i).Int("max_attempts", maxAttempts).Msg("failed to connected to the staked access node")
+			time.Sleep(time.Second)
+		}
+		// log fatal as there is no point continuing further if the unstaked AN cannot connect to the staked AN
+		log.Fatal().Msg("failed to connected to the staked access node, " +
+			"please ensure the node ID, network address and public key of the staked access node are correct " +
+			"and that the staked access node is running.")
+	})
+	return connector.lm.Started()
+}
+
+func (connector *upstreamConnector) Done() <-chan struct{} {
+	connector.lm.OnStop(func() {
+		if connector.cancel != nil {
+			connector.cancel()
+		}
```

Comment:
nit: if we create the context and store the cancelFunc in the constructor, we wouldn't need to check for `nil` here.

Reply:
agree - changed the constructor to accept a context and derive a cancel out of it.

Reply:
Actually, `LifecycleManager` guarantees that if `OnStop` is called before `OnStart`, then the shutdown function won't run. In other words, shutdown only runs if startup actually happens.

So in this case, we actually don't need to check for `nil` at all, because if we are executing the shutdown function then we know that startup occured (and hence that `connector.cancel` has been set. And at the same time, we can avoid creating the context in the constructor and having to store it inside a struct, which as mentioned previously should be avoided when possible.

Reply:
Actually, `LifecycleManager` guarantees that if `OnStop` is called before `OnStart`, then the shutdown function won't run. In other words, shutdown only runs if startup actually happens.

So in this case, we actually don't need to check for nil at all, because if we are executing the shutdown function then we know that startup occured (and hence that `connector.cancel` has been set. And at the same time, we can avoid creating the context in the constructor and having to store it inside a struct, which as mentioned previously should be avoided when possible.

Reply:
so the context is being stored so that it can be later passed to `AddPeer` when `Ready` is executed

Reply:
We can create the context in `Ready` in that case, and just store the `cancel`. We don't need to check for `nil` when `Done` is executed, because if it does execute at all then we know that `OnStart` has run and hence the `cancel` is not nil.

Reply:
to create the context, I will have to do a `context.Background()` in `Ready`. I wanted to use the context that was passed down by the builder.
Are you suggesting that I should delay the `cancel` creation until `Ready` is called?